File: python/get_images.py
from sentinelhub import (
    CRS,
    bbox_to_dimensions,
)
from sentinelhub import Geometry
import json
from datetime import datetime, timedelta
from request_image import request_image
import os
import json
from create_pixel_dataset import get_values, are_indexes_valid, get_indexes_from_image
import rasterio
from io import BytesIO

def get_images(geometry, farm, plot, start_date, end_date, satellite_interval_days = 5):

  polygon = Geometry(geometry, CRS.WGS84)
  betsiboka_size = bbox_to_dimensions(polygon.bbox, resolution=10)
  width_pixels = betsiboka_size[0]
  height_pixels = betsiboka_size[1]

  images_not_found_file = 'images/images_not_found.json'
  if os.path.exists(images_not_found_file):
      try:
          with open(images_not_found_file, 'r') as f:
              images_not_found = json.load(f)
      except json.JSONDecodeError as e:
          print(f"Warning: {images_not_found_file} contains invalid JSON. Initializing as an empty list.")
          raise e
  else:
      images_not_found = []
  
  images = {}
  
  # print(f"Getting images from {start_date.strftime('%Y-%m-%d')} to {end_date.strftime('%Y-%m-%d')}")
  while start_date <= end_date:
    start_date_str = start_date.strftime("%Y-%m-%dT00:00:00Z")
    end_date_str = start_date.strftime("%Y-%m-%dT23:59:59Z")
    try:
      file_name = f"images/{farm}_{plot}/{farm}_{start_date.strftime("%Y-%m-%d")}.tif"

      if file_name in images_not_found:
        # print(f"Image {file_name} is in the not found list. Skipping request.")
        start_date += timedelta(days=1) 
        continue

      if os.path.exists(file_name):
        # print(f"Image {file_name} already exists. Skipping request.")
        start_date += timedelta(days=satellite_interval_days)
        with open(file_name, "rb") as f:
          images[start_date_str.split('T')[0]] = BytesIO(f.read())
        continue
    
      image = request_image(start_date_str, end_date_str, geometry, width_pixels, height_pixels)

      with rasterio.open(BytesIO(image)) as dataset:
          x_range = range(0, dataset.width)
          y_range = range(0, dataset.height)
          total_pixels = dataset.width * dataset.height
      indexes = get_indexes_from_image(image)
      count = 0
      for y in y_range:
        for x in x_range:  
          ndmi_value, cld_value, scl_value, ndre_value, psri_value, b02_value, b04_value, ndvi_value = get_values(indexes,x,y)
          if not are_indexes_valid(psri_value, ndvi_value, ndmi_value, ndre_value, cld_value, scl_value, b02_value, b04_value):
              count += 1
      if count == total_pixels:
        raise Exception("Image not found")

      if not os.path.exists('images'):
        os.makedirs('images')
      if not os.path.exists(f'images/{farm}_{plot}'):
        os.makedirs(f'images/{farm}_{plot}')
      with open(file_name, "wb") as f:
        f.write(image)
        images[start_date_str.split('T')[0]] = BytesIO(image)
        
      # print(f"Image saved as {file_name}")
    except Exception as e:
      if str(e) == "Image not found":
        # print(f"No image found for date {start_date.strftime('%Y-%m-%')}")
        images_not_found.append(file_name)
        with open(images_not_found_file, 'w') as f:
          json.dump(images_not_found, f)
      else:
        print(f"Error: {e}")
      start_date += timedelta(days=1)
      continue
    start_date += timedelta(days=satellite_interval_days)

  return images

  

File: go/get_images.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"time"

	"github.com/lukeroth/gdal"
)

// GetImages retrieves satellite images based on the given parameters
func GetImages(geometry map[string]any, farm, plot string, startDate, endDate time.Time, satelliteIntervalDays int) (map[string]gdal.Dataset, error) {
	images := make(map[string]gdal.Dataset)
	imagesNotFoundFile := "images/images_not_found.json"

	// Load images_not_found.json
	var imagesNotFound []string
	if _, err := os.Stat(imagesNotFoundFile); err == nil {
		data, err := ioutil.ReadFile(imagesNotFoundFile)
		if err != nil {
			return nil, fmt.Errorf("failed to read %s: %v", imagesNotFoundFile, err)
		}
		if err := json.Unmarshal(data, &imagesNotFound); err != nil {
			return nil, fmt.Errorf("invalid JSON in %s: %v", imagesNotFoundFile, err)
		}
	}

	// Ensure images directory exists
	if _, err := os.Stat("images"); os.IsNotExist(err) {
		if err := os.Mkdir("images", os.ModePerm); err != nil {
			return nil, fmt.Errorf("failed to create images directory: %v", err)
		}
	}

	// Iterate through dates
	for currentDate := startDate; !currentDate.After(endDate); currentDate = currentDate.AddDate(0, 0, satelliteIntervalDays) {
		startDateStr := currentDate.Format("2006-01-02T00:00:00Z")
		endDateStr := currentDate.Format("2006-01-02T23:59:59Z")
		fileName := fmt.Sprintf("images/%s_%s/%s_%s.tif", farm, plot, farm, currentDate.Format("2006-01-02"))

		// Skip if image is in the not-found list
		if contains(imagesNotFound, fileName) {
			continue
		}

		// Skip if file already exists
		if _, err := os.Stat(fileName); err == nil {
			data, err := gdal.Open(fileName, gdal.Access(gdal.ReadOnly))
			if err != nil {
				return nil, fmt.Errorf("failed to open %s: %v", fileName, err)
			}
			images[currentDate.Format("2006-01-02")] = data
			continue
		}

		// Request image
		image, err := requestImage(startDateStr, endDateStr, geometry, 0, 0) // Width and height are placeholders
		if err != nil {
			if err.Error() == "Image not found" {
				imagesNotFound = append(imagesNotFound, fileName)
				saveImagesNotFound(imagesNotFoundFile, imagesNotFound)
				continue
			}
			return nil, fmt.Errorf("error requesting image: %v", err)
		}

		// Validate image (mocked logic)
		totalPixels := 100 // Placeholder for total pixels
		count := 0
		for y := 0; y < 10; y++ { // Placeholder for height
			for x := 0; x < 10; x++ { // Placeholder for width
				_, _, _, _, _, _, _, _ = getValues(nil, x, y)
				if !areIndexesValid(0, 0, 0, 0, 0, 0, 0, 0) {
					count++
				}
			}
		}
		if count == totalPixels {
			imagesNotFound = append(imagesNotFound, fileName)
			saveImagesNotFound(imagesNotFoundFile, imagesNotFound)
			continue
		}

		// Save image to file
		dirPath := filepath.Dir(fileName)
		if _, err := os.Stat(dirPath); os.IsNotExist(err) {
			if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
				return nil, fmt.Errorf("failed to create directory %s: %v", dirPath, err)
			}
		}
		if err := ioutil.WriteFile(fileName, image, 0644); err != nil {
			return nil, fmt.Errorf("failed to save image to %s: %v", fileName, err)
		}
		data, err := gdal.Open(fileName, gdal.Access(gdal.ReadOnly))
		if err != nil {
			return nil, fmt.Errorf("failed to open %s: %v", fileName, err)
		}

		images[currentDate.Format("2006-01-02")] = data
	}

	return images, nil
}

func saveImagesNotFound(filePath string, imagesNotFound []string) {
	data, _ := json.Marshal(imagesNotFound)
	_ = os.WriteFile(filePath, data, 0644)
}

func contains(slice []string, item string) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

// func main() {
// 	// Example usage
// 	geometry := Geometry{Coordinates: nil, CRS: "WGS84"}
// 	startDate, _ := time.Parse("2006-01-02", "2023-01-01")
// 	endDate, _ := time.Parse("2006-01-02", "2023-01-10")
// 	images, err := GetImages(geometry, "farm1", "plot1", startDate, endDate, 5)
// 	if err != nil {
// 		fmt.Println("Error:", err)
// 		return
// 	}
// 	fmt.Println("Images retrieved:", len(images))
// }
